sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala

```diff
@@ -896,6 +896,19 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
     }
   }
 
+  test("SPARK-24012 Union of map and other compatible columns") {
+    checkAnswer(
+      sql(
+        """
+          |SELECT map(1, 2), 'str'
+          |UNION ALL
+          |SELECT map(1, 2, 3, NULL), 1""".stripMargin),
```

Comment:
can you give some insight about why it doesn't work? I'd expect Spark first do type coercion for `map(1, 2, 3, NULL)`, and the result is map<int, nullable int>, then Union should accept nullability difference and pass the analysis.

Reply:
map<int, nullable int> and map<int, not nullable int> are accepted by Union, but, string and int are not.

If types of one column can not be accepted by Union, TCWSOT(TypeCoercion.WidenSetOperationTypes) will try to coerce them to a completely identical type. TCWSOT works when all of the columns can be coerced and not work when columns can not be coerced exist.
map<int, nullable int> and map<int, not nullable int> can not be coerced, so, TCWSOT didn't work and string and int will not be coerced.

Reply:
Shall we make map<int, nullable int> and map<int, not nullable int> coerce-able?

Reply:
Of course we can. 
two solution：
1. Try cast two map types to one no matter key types are not the same or value types are not the same.
`select map(1, 2) union all map(1, 'str')` will work.
2. Cast two map type to one only when the key type and value type are the same. This solution just resolve the problem that map<t1, nullable t2> and map<t1, not nullable t2> can't be union.